performance_test/src/communication_abstractions/opendds_communicator.hpp

```diff
@@ -0,0 +1,344 @@
+  explicit OpenDDSCommunicator(SpinLock & lock)
+  : Communicator(lock),
+    m_datawriter(nullptr),
+    m_datareader(nullptr),
+    m_typed_datareader(nullptr)
+  {
+
+    m_participant = ResourceManager::get().opendds_participant();
+    register_topic();
+  }
+
+  /**
+   * \brief Publishes the provided data.
+   *
+   *  The first time this function is called it also creates the data writer.
+   *  Further it updates all internal counters while running.
+   * \param data The data to publish.
+   * \param time The time to fill into the data field.
+   */
+  void publish(DataType & data, const std::chrono::nanoseconds time)
+  {
+    if (m_datawriter == nullptr) {
+      DDS::Publisher_ptr publisher;
+      DDS::DataWriterQos dw_qos;
+      ResourceManager::get().opendds_publisher(publisher, dw_qos);
+
+      dw_qos.resource_limits.max_samples = 32;
+      dw_qos.resource_limits.max_samples_per_instance = 32;
+      dw_qos.resource_limits.max_instances = 1;
+
+      OpenDdsQOSAdapter qos_adapter(m_ec.qos());
+      qos_adapter.apply_dw(dw_qos);
+
+      m_datawriter = publisher->create_datawriter(m_topic,
+          dw_qos, nullptr, OpenDDS::DCPS::DEFAULT_STATUS_MASK);
+      if (CORBA::is_nil(m_datawriter)) {
+        throw std::runtime_error("Could not create datawriter");
+      }
+
+      m_typed_datawriter = DataWriterType::_narrow(m_datawriter);
+      if (CORBA::is_nil(m_typed_datawriter)) {
+        throw std::runtime_error("failed datawriter narrow");
+      }
+    }
+    lock();
+    data.time_ = time.count();
+    data.id_ = next_sample_id();
+    increment_sent();  // We increment before publishing so we don't have to lock twice.
+    unlock();
+    auto retcode = m_typed_datawriter->write(data, DDS::HANDLE_NIL);
+    if (retcode != DDS::RETCODE_OK) {
+      throw std::runtime_error("Failed to write to sample");
+    }
+  }
+
+  /**
+   * \brief Reads received data from DDS.
+   *
+   * In detail this function:
+   * * Reads samples from DDS.
+   * * Verifies that the data arrived in the right order, chronologically and also consistent with the publishing order.
+   * * Counts received and lost samples.
+   * * Calculates the latency of the samples received and updates the statistics accordingly.
+   */
+
+  void update_subscription()
+  {
+    if (CORBA::is_nil(m_datareader)) {
+      DDS::Subscriber_ptr subscriber = nullptr;
+      DDS::DataReaderQos dr_qos;
+      ResourceManager::get().opendds_subscriber(subscriber, dr_qos);
+
+      dr_qos.resource_limits.max_samples = 32;
+      dr_qos.resource_limits.max_instances = 10;
+      dr_qos.resource_limits.max_samples_per_instance = 32;
+      /* if there are more remote writers, you need to increase these limits */
+      
+      //These are not available in OpenDDS, RTI tricks
+      //dr_qos.reader_resource_limits.max_remote_writers = 10;
+      //dr_qos.reader_resource_limits.max_remote_writers_per_instance = 10;
```

Comment:
If this isn't supported in OpenDDS, you can just remove it

Reply:
done